src/inmanta/server/services/userservice.py

```diff
@@ -0,0 +1,125 @@
+"""
+    Copyright 2023 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+import logging
+
+import asyncpg
+
+import nacl.exceptions
+import nacl.pwhash
+from inmanta import const, data, protocol
+from inmanta.const import MIN_PASSWORD_LENGTH
+from inmanta.data import AuthMethod, model
+from inmanta.protocol import common, exceptions
+from inmanta.server import SLICE_DATABASE, SLICE_TRANSPORT, SLICE_USER
+from inmanta.server import config as server_config
+from inmanta.server import protocol as server_protocol
+
+LOGGER = logging.getLogger(__name__)
+
+
+def verify_authentication_enabled() -> None:
+    """raises an UnauthorizedException exception if server authentication is not enabled"""
+    if not server_config.server_enable_auth.get():
+        raise exceptions.BadRequest(
+            "Server authentication should be enabled. To setup the initial user use the inmanta-initial-user-setup tool."
+        )
+
+
+class UserService(server_protocol.ServerSlice):
+    """Slice for managing users"""
+
+    def __init__(self) -> None:
+        super().__init__(SLICE_USER)
+
+    def get_dependencies(self) -> list[str]:
+        return [SLICE_DATABASE]
+
+    def get_depended_by(self) -> list[str]:
+        return [SLICE_TRANSPORT]
+
+    @protocol.handle(protocol.methods_v2.list_users)
+    async def list_users(self) -> list[model.User]:
+        return sorted([user.to_dao() for user in await data.User.get_list()], key=lambda u: u.username)
+
+    @protocol.handle(protocol.methods_v2.add_user)
+    async def add_user(self, username: str, password: str) -> model.User:
```

Comment:
Shouldn't we also make sure that the username is not the empty string? I guess that having an empty string as a username can cause all sorts of issues. We should also validate this in the user setup script.

Reply:
In the user setup script, if no username is provided it defaults to Admin. So I don't think a check is needed there. But I've added it to the test